File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head == None:
            return False 
       
        p1 = head
        p2 = head

        while True:
            p1 = p1.next

            p2 = p2.next

            if p2 == None:
                return False

            p2 = p2.next

            if p2 == None:
                return False
        
            if p1 is p2:
                return True

File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func reverseList(head *ListNode) *ListNode {
    h := head;

    if h == nil {
        return h
    }

    m := h.Next

    if m == nil {
        return h
    }

    h = nil
    m = head

    t := m.Next

    for {
        m.Next = h

        h = m
        m = t

        if t == nil {
            break
        }

        t = t.Next
    }

    return h
}
